### ISSUE TITLE
DNS broken - /mnt/wsl is empty - resolv.conf missing

### ISSUE THREAD

**Issue by torgeros** [microsoft/WSL#11928]
**NOTE**: This might be a reopen of #11036 

### Windows Version

Microsoft Windows [Version 10.0.22631.3880]

### WSL Version

2.2.4.0

### Are you using WSL 1 or WSL 2?

- [X] WSL 2
- [ ] WSL 1

### Kernel Version

5.15.153.1-microsoft-standard-WSL2

### Distro Version

Ubuntu 22.04 LTS

### Other Software

Docker is installed (WSL-based backend)

### Repro Steps

I don't know

### Expected Behavior

accessing the internet being possible inside wsl

### Actual Behavior

Every hostname- / url-based network request fails with `Temporary failure in name resolution`. The internet guided me towards the `/etc/resolv.conf` file. This file, however, is just a symlink to `/mnt/wsl/resolv.conf` which, from what I have read, should be automatically generated every time WSL boots. However, the `/mnt/wsl` directory is completely empty. When I create a file inside `/mnt/wsl`, it is deleted after a WSL reboot.

And obviously, with `/etc/wsl/resolv.conf` not existing, my DNS cannot resolve anything.

WSL used to work previously, I do not know when exactly it stopped working.

### Diagnostic Logs

/etc/wsl.conf:

```plain

[boot]
systemd=true
```

**Comment by torgeros** [comment]
This is a duplicate of #11693.

**Comment by torgeros** [comment]
As it turned out in https://github.com/microsoft/WSL/issues/11693#issuecomment-2296475546, these issues are not the same. I am completely missing my wsl config whereas theirs is at least present...

**Comment by torgeros** [comment]
Fyi, I cannot take the logs because I do not have administrator-level access.

**Comment by OneBlue** [comment]
Unfortunately logs are required for us to being able to root cause issue. I'll close for now but feel free to reopen if you can collect logs.

**Comment by kohlerdominik** [comment]
Hi @torgeros @OneBlue 

I have this issue and already posted in #11693.
Here are my logs: https://github.com/user-attachments/files/15840532/WslLogs-2024-06-14_17-43-54.zip
Please reopen.

@torgeros Do you have a LAN having a networkadress in the 172.x.x.x. range?

@OneBlue In my case this issue appeared after a couple of days having my new notebook when I connected it to the company network `172.25.1.0/16`. 
I think the problem originates from WSL assigning a random 172.x.x.x-network for the WSL environment. So in that case, it Auto-Assigned some `172.25.?.0/20` network, and I joined an overlapping network, it crashed. My network `/mnt/wsl/resolv.conf` is missing ever since.

**Comment by torgeros** [comment]
> @torgeros Do you have a LAN having a networkadress in the 172.x.x.x. range?

The only thing in my `ipconfig` with a 172 address is the WSL adapter.

**Comment by oKazuhiro** [comment]
Hi @kohlerdominik the issue you have seems very similar to the issue I am having.

I am experiencing a similar issue where WSL is unable to generate and link the `resolve.conf` file. My host's Ethernet adapter is using a static IP in range `172.x.x.x`. I actually have Hyper-V disabled, but Virtual Machine Platform is enabled so that I can use WSL2.

I was using mirrored networking mode, but switched back to the default NAT networking mode which fixed the connectivity issues, but I need to use mirrored networking mode due to some issues with port forwarding.

**Comment by kohlerdominik** [comment]
Hi @oKazuhiro 

Can you check wether `/mnt/wsl/resolv.conf` exists in your WSL-Distro?
![grafik](https://github.com/user-attachments/assets/87c7ab6f-7953-4cba-80ee-608acaf51b1b)

**Comment by oKazuhiro** [comment]
> Hi @oKazuhiro
> 
> Can you check wether `/mnt/wsl/resolv.conf` exists in your WSL-Distro? ![grafik](https://private-user-images.githubusercontent.com/18621527/368057839-87c7ab6f-7953-4cba-80ee-608acaf51b1b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjY2OTM4NjQsIm5iZiI6MTcyNjY5MzU2NCwicGF0aCI6Ii8xODYyMTUyNy8zNjgwNTc4MzktODdjN2FiNmYtNzk1My00Y2JhLTgwZWUtNjA4YWNhZjUxYjFiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTE4VDIxMDYwNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTZjOWE1YjY3YTVkNGZjYTk2OTdjOTNkM2ZhMzYwMTdmMDAwZGEzNWYxZGJhMWRhMDM1NWQ3ZmY5MTY5NjczYjYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.U3Ui-iXXKMAHKSNM78IsQ0uNJ3JU-9EFfbnrAy9EeHY)

The symbolic link does exist at `/etc/resolv.conf`, but the file it links to in `/mnt/wsl/resolv.conf` is missing.

**Comment by kohlerdominik** [comment]
@torgeros Thanks. 

How can we help with fixing the issue? So it seems that we broke the generation of the resolv.conf, presumably by forcing WSL to switch to another IP-address for its virtual network (`Ethernet-Adapter vEthernet (WSL (Hyper-V firewall))`), because the one it choose initialize was taken by a physical network adapter.

**Comment by oKazuhiro** [comment]
Found a solution that worked for me.

I ended up having to enable IPv6 in the registry.

This issue helped me come to this solution: #11002

1. Open regedit.
2. Navigate to HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Services\Tcpip6\Parameters.
3. Modify DisabledComponents to 20 (Hex) to prefer IPv4 while keeping IPv6 enabled. or 00 (Hex) to fully enable IPv6.
4. Restart the system.

**Comment by kohlerdominik** [comment]
@oKazuhiro So with your fix, the `/mnt/wsl/resolv.conf` is back again? Unfortunatelly, for me it didn't do anything...

**Comment by torgeros** [comment]
The `DisabledComponents` key does not even exist in my registry. resolv.conf generation, however, seem to be permanently broken because across system reboots, even without any networks connected, it does not generate that file.

**Comment by kohlerdominik** [comment]
> I was able to solve mine by setting
> 
> ```
> [wsl2]
> dnsTunneling=false
> ```
> 
> in `%UserProfile%\.wslconfig`.

Woulnd't that bypass host entries on the Windows instance? Meaning, this is more a bandaid than a solution?

**Comment by torgeros** [comment]
@kohlerdominik Basically, per [the docs](https://learn.microsoft.com/en-us/windows/wsl/networking#dns-tunneling), `dnsTunneling` specifies the opposite of what it sounds like.  
With `dnsTunneling=true`, dns requests are processed inside WSL by, and I am quoting here, "a virtualization feature" which is not further specified.  
However, with `dnsTunneling=false`, dns requests are actually passed through as networking packets and resolved on the windows host.

And yes, my /mnt/wsl/resolv.conf is back with the following content:

```conf
# This file was automatically generated by WSL. To stop automatic generation of this file, add the following entry to /etc/wsl.conf:
# [network]
# generateResolvConf = false
nameserver 172.19.0.1
```

This matches with `172.19.0.1` as my windows host IP (`ip route show | grep -i default | awk '{ print $3}'`)

**Comment by kohlerdominik** [comment]
@torgeros Thank you very much, this indeed fixed it for me as well. Still no idea how it broke for `dnsTunneling=true`, though...

**Comment by CidTori** [comment]
> I was able to solve mine by setting
> 
> ```
> [wsl2]
> dnsTunneling=false
> ```
> 
> in `%UserProfile%\.wslconfig`.

This does not fix anything in 2.5.9.0

**Comment by CidTori** [comment]
> Just had this issue of resolv.conf missing and the symlink pointing to a non existent resolv.conf in wsl dir. Rebooting wsl did not generate a new resolv.conf. My wslconfig had:
> 
> ```
> $ cat /etc/wsl.conf
> # https://learn.microsoft.com/en-us/windows/wsl/wsl-config#example-wslconf-file
> [boot]
> systemd=true
> command = ntpdate time.windows.com; service cron start; service docker start'
> ```
> 
> Added explicit params to config but that made no difference:
> 
> ```
> [network]
> generateHosts = true
> generateResolvConf = true
> ```
> 
> I created `.wslconfig` in windows as there was none (read docs about it) and set:
> 
> ```
> [wsl2]
> debugConsole=true
> ```
> 
> Then:
> 
>     * restarted wsl with `wsl.exe --shutdown`
> 
>     * after starting debug window appeared but it got stuck on login and could not proceed.
> 
>     * updated Windows and restarted it
> 
>     * started wsl again and this time the debug window was not stuck on login and wsl instance successfully booted
> 
>     * resolv.conf was generated
> 
> 
> ```
> $ ls -la /etc/resolv.conf
> /etc/resolv.conf -> /mnt/wsl/resolv.conf
> $ ls /mnt/wsl/resolv.conf
> /mnt/wsl/resolv.conf
> $ cat /mnt/wsl/resolv.conf
> # This file was automatically generated by WSL. To stop automatic generation of this file, add the following entry to /etc/wsl.conf:
> # [network]
> # generateResolvConf = false
> nameserver 10.255.255.254
> ```

I couldn’t replicate this
